### PR TITLE
Refactor event registration API

### LINKED ISSUE #12: Refactor Event Register API
**Describe the issue**
We are improving the reliability and reusability of our code. Thereby, we need to simplify and document our code. Use good DRF practices taking <a href="https://github.com/ECellNitrr/teesco-backend">teesco-backend</a> as reference.

Navigate to `events/views.py` and refer `event_register` view and simplify it. Also write its swagger documentation.

**Expected behaviour**
Code must be understandable and reproducible.

### PR DESCRIPTION
**Pull request by KunalAnkur**
### Description
Event Registration API has been updated into class-based view and removed all unnecessary imports.

**Type of Change:** (Delete irrelevant options)
- Code

**Code/Quality Assurance Only** (Delete irrelevant options)

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?
It has been tested on locally.

### Checklist:
- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [ ] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] Any dependent changes have been merged
- [x] Update Swagger documentation and the exported file at /docs folder
- [ ] Update requirements.txt

**Code/Quality Assurance**
- [x] My changes generate no new warnings 
- [ ] My PR currently breaks something (fix or feature that would cause existing functionality to not work as expected)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been published in downstream modules

### CODE REVIEW

**Comment by priyanshisharma** [comment]
@KunalAnkur please pull from upstream and resolve conflicts.

**Comment by KunalAnkur** [comment]
> @KunalAnkur please pull from upstream and resolve conflicts.

done ma'am

File: events/responses.py [hunk 1]
```diff
@@ -23,4 +23,16 @@
...
+event_registration_200 = {
+    "message":"Registration Successful"
+}
+
+event_does_not_exist_404 = {
+  "message": "Registration Failed. Event does not exist."
+}
+
+user_unauthorized_401 = {
+  "message":"Unauthorized User"
```
Comment (priyanshisharma) [inline events/responses.py:37]:
```suggestion
  "message":"Please follow OTP verification before registering"
```
It is good to mention 'why?' when we can in the, response message.

File: events/views.py [hunk 2]
```diff
@@ -30,32 +31,31 @@ def get(self,request,year):
...
+        if user.verified:
+            eventregister.user = user
+            try:
+                eventregister.event = Event.objects.get(id=id)
+            except:
+                return Response(responses.event_does_not_exist_404, status.HTTP_404_NOT_FOUND)    
+            else:
+                eventregister.save()
+                return Response(responses.event_registration_200,status.HTTP_200_OK)
+        else:
```
Comment (priyanshisharma) [inline events/views.py:57]: This `else` is redundant and can be eliminated

File: events/views.py [hunk 3]
```diff
@@ -30,32 +31,31 @@ def get(self,request,year):
...
+    def post(self,request,id):
+        user = request.user
+        eventregister = EventRegister()
+        if user.verified:
+            eventregister.user = user
+            try:
+                eventregister.event = Event.objects.get(id=id)
+            except:
+                return Response(responses.event_does_not_exist_404, status.HTTP_404_NOT_FOUND)    
+            else:
```
Comment (priyanshisharma) [inline events/views.py:54]: This `else` is redundant and can be eliminated.

**Review by priyanshisharma — CHANGES_REQUESTED** [review]
Looks Nice 🙌 
Observe the comments given and Add a condition to check if the user is already registered for an event, else it fails the `UNIQUE constraint` and gives `IntegrityError`.

File: events/views.py (hunk 3 above)
Comment (KunalAnkur) [reply events/views.py]: done

File: events/views.py (hunk 2 above)
Comment (KunalAnkur) [reply events/views.py:63]: done

File: events/responses.py (hunk 1 above)
Comment (KunalAnkur) [reply events/responses.py]: done

File: events/views.py [hunk 4]
```diff
@@ -37,25 +37,31 @@ class EventRegisterView(APIView):
...
-        if user.verified:
-            eventregister.user = user
-            try:
-                eventregister.event = Event.objects.get(id=id)
-            except:
-                return Response(responses.event_does_not_exist_404, status.HTTP_404_NOT_FOUND)    
-            else:
+        try:
+            event = Event.objects.get(id=id)
+            registeredUser = EventRegister.objects.get(user=user,event=event)
```
Comment (priyanshisharma) [inline events/views.py:51]: First use a `try` and `except` statement for finding the `event` where you respond with event not found if the event isn't found. Do not mix it with finding registered user, so it after the `except` response.

File: events/views.py [hunk 5]
```diff
@@ -37,25 +37,31 @@ class EventRegisterView(APIView):
...
+        except:
+            if user:
+                eventregister.user = user
+                try:
+                    eventregister.event = Event.objects.get(id=id) 
+                except:
+                    return Response(responses.event_does_not_exist_404, status.HTTP_404_NOT_FOUND)
```
Comment (priyanshisharma) [inline events/views.py:58]: You are already looking for the even in line 50, then checking here again, this is generating unnecessary work.

File: events/views.py [hunk 6]
```diff
@@ -37,25 +37,31 @@ class EventRegisterView(APIView):
...
-            try:
-                eventregister.event = Event.objects.get(id=id)
-            except:
-                return Response(responses.event_does_not_exist_404, status.HTTP_404_NOT_FOUND)    
-            else:
+        try:
+            event = Event.objects.get(id=id)
+            registeredUser = EventRegister.objects.get(user=user,event=event)
+        except:
+            if user:
```
Comment (priyanshisharma) [inline events/views.py]:
```suggestion
            if user.verified:
```
Check if the user is verified.

File: events/views.py (hunk 2 above)
Comment (priyanshisharma) [reply events/views.py:63]: Isn't the else that has now shifted to line number 63 still redundant? 🤔

**Review by priyanshisharma — CHANGES_REQUESTED** [review]
I see you checked whether the user is registered in the event first, even before knowing whether the event itself exists, or the user is verified. In the comments I've given, observe how this makes the code a little unpredictable. Look the points below and try to follow this chain of thought:

1. Does the event itself exist? If no, there is nowhere to register.
2. Now that the event exists, is the user eligible to register (is s/he verified)?
3. Now that s/he is verified, is the user already registered in the event s/he wants to register in? If yes, then we're good, nothing left on our part.
4. If not, Lets register them!

I would recommend rewriting the code, keeping this in mind. Use the code you've already written as reference.

File: events/views.py (hunk 4 above)
Comment (KunalAnkur) [reply events/views.py]: okay ma'm

File: events/views.py (hunk 5 above)
Comment (KunalAnkur) [reply events/views.py]: done

File: events/views.py (hunk 6 above)
Comment (KunalAnkur) [reply events/views.py]: done

**Review by priyanshisharma — APPROVED** [review]
Looks good to me, good work! 🎉

File: events/views.py [hunk 7]
```diff
@@ -30,32 +31,41 @@ def get(self,request,year):
...
+            '201': set_example(responses.event_registration_201),
+            '200':set_example(responses.user_already_registered_event_200),
+            '404':set_example(responses.event_does_not_exist_404),
+            '401':set_example(responses.user_unauthorized_401)
+        }
+    )
+    def post(self,request,id):
+        user = request.user
+        eventregister = EventRegister()
+        event = Event.objects.all()
```
Comment (anshsrtv) [inline events/views.py:49]: I did not understand why this is here. Is there a necessity?
```suggestion
```

**Review by anshsrtv — CHANGES_REQUESTED** [review]
Well written code 🚀 

Kudos to both of you 🔥 @KunalAnkur @priyanshisharma 🙇

File: events/views.py (hunk 7 above)
Comment (KunalAnkur) [reply events/views.py]: done sir

**Review by anshsrtv — APPROVED** [review]

**Pull request merged**